### PR TITLE
Create ServiceMonitors to gather metrics from the OTEL Operands

### LINKED ISSUE #1768: Support collecting metrics from instances
The collector instances deployed by the OpenTelemetry Operator expose internal metrics. We can collect those metrics to get more information about how our system is behaving.

[I wrote this proposal to implement the feature](https://docs.google.com/document/d/1bUiwgnkN3rI0Z_vm9_cAECdCVtqtNN8QQ4UJcnenKxE/edit?usp=sharing).

### PR DESCRIPTION
**Pull request by iblancasa**
Resolves #1768

### CODE REVIEW

File: pkg/collector/reconcile/servicemonitor.go
Comment (TylerHelmuth) [inline pkg/collector/reconcile/servicemonitor.go:57]: I don't love the idea of the OpenTelemetry Operator being able to create objects that depend on other CRDs being preset, it feels like an opportunity for our users to get into weird situations.

Comment (jaronoff97) [reply pkg/collector/reconcile/servicemonitor.go:57]: this also creates a direct dependency on the prometheus operator from ours which concerns me. Right now we have a dependency in the TA which makes sense as it is the component pulling down that functionality, but I worry what happens with this added dependency.

Comment (iblancasa) [reply pkg/collector/reconcile/servicemonitor.go:57]: > I don't love the idea of the OpenTelemetry Operator being able to create objects that depend on other CRDs being preset

From the conversations we had during the 25 May SIG meeting, I understood it was OK to implement the proposal described in #1768.

> it feels like an opportunity for our users to get into weird situations.

What situations? I would be happy to check them and apply a fix if needed.

> I agree, I think this opens up some issues if we make the assumption that every operator user wants to use service monitors. 

I don't think this will be a problem since everything is protected with a feature flag. If the feature flag is not enabled, the Service Monitors will not be created. Also, you need to set to true the feature in the OpenTelemetry Collector instance.

> This is a good topic for our next SIG meeting.

I'll not be able to attend since I'll be on vacation next week. But I'll be happy to talk about it in another moment.

Comment (pavolloffay) [reply pkg/collector/reconcile/servicemonitor.go:57]: We have discussed this on the SIG meeting that the operator will create Service/Pod monitors for operands with the assumption that the functionality will be put behind a feature flag. The implementation has to be flexible to allow OTLP in the future (once the collector will report OTLP metrics).

This work is needed to move the OTEL operator to operator level 4 https://operatorhub.io/operator/opentelemetry-operator. There is a lot of other operators that do the same.

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (pavolloffay) [inline apis/v1alpha1/opentelemetrycollector_types.go:357]: Document here, how this feature is enabled

File: pkg/collector/reconcile/servicemonitor.go
Comment (jaronoff97) [reply pkg/collector/reconcile/servicemonitor.go:57]: ah okay, yeah sorry checked my notes and saw this now. Proceed :) will review now.

File: main.go
Comment (jaronoff97) [inline main.go:45]: this should be grouped with the above.

File: main.go
Comment (jaronoff97) [inline main.go:76]: Should we also put this behind the flag?

**Review by jaronoff97 — APPROVED** [review]
The approach here makes sense, could you also add an e2e test please?

File: main.go
Comment (iblancasa) [reply main.go:76]: When this code is executed, the flag value is unknown.

File: tests/e2e/prometheus-config-validation/03-promreceiver-nopromconfig.yaml
Comment (jaronoff97) [inline tests/e2e/prometheus-config-validation/03-promreceiver-nopromconfig.yaml:10]: Shouldn't we be using the local image here?

Comment (iblancasa) [reply tests/e2e/prometheus-config-validation/03-promreceiver-nopromconfig.yaml:10]: Sorry. I added changes that should not be in the PR. Fixing...

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (TylerHelmuth) [inline apis/v1alpha1/opentelemetrycollector_types.go:384]: Is `CreateServiceMonitors` generic enough to handle the switch to the Collector when the time comes?  It would be nice if our Config API doesn't need any breaking changes when we can replace Prometheus with the Collector's exporting capabilities.

File: controllers/opentelemetrycollector_controller.go
Comment (TylerHelmuth) [inline controllers/opentelemetrycollector_controller.go:173]: Will this task cause problems when the feature gate is `false`?

File: pkg/collector/reconcile/servicemonitor.go
Comment (TylerHelmuth) [inline pkg/collector/reconcile/servicemonitor.go:70]: Where is this port being defined?  Does this handle situations where someone manually configures the collector's metrics port to be something other than the default?

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (iblancasa) [reply apis/v1alpha1/opentelemetrycollector_types.go:384]: Ok I agree the name is not as generic as it should be. How about `EnableMetrics`?

File: controllers/opentelemetrycollector_controller.go
Comment (iblancasa) [reply controllers/opentelemetrycollector_controller.go:173]: The first thing to be checked during that task is if the feature flag is enabled:
```go
if !params.Instance.Spec.Observability.Metrics.CreateServiceMonitors || !featuregate.PrometheusOperatorIsAvailable.IsEnabled() {
```

File: pkg/collector/reconcile/servicemonitor.go
Comment (iblancasa) [reply pkg/collector/reconcile/servicemonitor.go:70]: It is defined here:
https://github.com/open-telemetry/opentelemetry-operator/blob/56f5de94dd3f1f83ac7dbdba3b8b4a6379317b48/pkg/collector/reconcile/service.go#L197-L200

It cannot be modified (right now). But I think it could be an interesting follow-up task: to allow the personalization of the metrics port. What do you t hink?

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (TylerHelmuth) [reply apis/v1alpha1/opentelemetrycollector_types.go:384]: I like that, and that will flex well into any metric generation solution.

File: pkg/collector/reconcile/servicemonitor.go
Comment (iblancasa) [reply pkg/collector/reconcile/servicemonitor.go:70]: Ok, I see your point.
Yes, the metrics become unreachable. If we want to address that use case, I would do it as part of a different issue since it would involve changes in https://github.com/open-telemetry/opentelemetry-operator/blob/56f5de94dd3f1f83ac7dbdba3b8b4a6379317b48/pkg/collector/reconcile/service.go#L197-L200

And look for a way to get the address where the metrics are exposed. What do you think?

Comment (TylerHelmuth) [reply pkg/collector/reconcile/servicemonitor.go:70]: Doing it as part of another issue is good.  I kinda feel like we need to solve that issue before we introduce this functionality.

Comment (iblancasa) [reply pkg/collector/reconcile/servicemonitor.go:70]: Created #1931

**Comment by jaronoff97** [comment]
@TylerHelmuth mind giving one more review?

**Comment by TylerHelmuth** [comment]
@jaronoff97 I am still concerned about https://github.com/open-telemetry/opentelemetry-operator/pull/1874#discussion_r1264034952. It is bad  that users can't modify their metrics endpoint today, I am hesitant to add on more functionality dependent on the metrics endpoint without the flexibility to configure it.  While I recognize we could release this first and add support for configuration later, it feels like an implementation that drives the Operator to Level 4 should also allow the users to configure the managed object (in this instance the Collector) however they need.

File: pkg/featuregate/featuregate.go
Comment (TylerHelmuth) [inline pkg/featuregate/featuregate.go:64]: Please add a `featuregate.FromVersion` that marks the Operator version this feature gate was added. (This was a rather late concept in the featuregate package and we should go back and fix our other gates to include it)

**Comment by jaronoff97** [comment]
I see, in that case let's block merging this until we have #1931 closed

**Comment by iblancasa** [comment]
> I see, in that case let's block merging this until we have #1931 closed

Even without merging this PR, the problem is still there because the service created by the operator will be pointing to a port that does nothing. I'll work on #1931 but, TBH, I don't think there is a real need to block this PR until #1931 is merged.

File: .github/workflows/e2e.yaml
Comment (TylerHelmuth) [inline .github/workflows/e2e.yaml:31]: Can we separate `e2e-prometheuscr` into its own section in the matrix?  `e2e e2e-upgrade` already has a noisy neighbor problem and I'd prefer not to add another.

Comment (iblancasa) [reply .github/workflows/e2e.yaml:31]: Sure. Just didn't want to separate in case somebody though it was not ok to create a new section.

Comment (TylerHelmuth) [reply .github/workflows/e2e.yaml]: Luckily we have the CI checks setup to look at the dependent job, so we can now scale this worker job as needed.

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**
